### PR TITLE
ADR for front-end code quality tools

### LINKED ISSUE #102: ADR: Front-end Code Quality Tools
### Description

Evaluate and select libraries to enforce the following standards within the frontend codebase:

- Dependency management (e.g. conflicts, updates, etc.)
- Code Linting
- Auto-formatting
- Type checking
- Checking for vulnerable dependencies
- Checking for open source licenses

### Definition of Done

- [ ] The approvers for this decision have been identified (ideally before work on the ADR starts)
- [ ] The ADR is created and stored in documentation/decisions/adr with status "Accepted"
- [ ] The ADR has been reviewed and approved by the approvers listed above
- [ ] The ADR satisfies requirements that are outlined in the ADR template
- [ ] Any follow-up tickets have been created (if necessary)
- [ ] The ADR covers each of the following tools:
  - [ ] Dependency management
  - [ ] Code linting
  - [ ] Auto-formatting
  - [ ] Type checking
  - [ ] Dependency vulnerability checking
  - [ ] License checking

### PR DESCRIPTION
**Pull request by daphnegold**
## Summary
Fixes #102

## Context for reviewers
Describes decision for front-end code quality tooling. Assumes we are using the Next.js template repository, however, set of tooling is extensible to any solution we choose.

### CODE REVIEW

File: documentation/decisions/adr/000N-frontend-tooling.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+## Decision Drivers <!-- RECOMMENDED -->
+
+- **Ease of use and configurability:** Code quality tools should be configurable to meet the specific needs of our project and enforce standards uniformly. Running the tool should be simple, and feedback should be easily implementable and understandable.
+- **Speed:** Code quality tools should be able to run in an timely manner, ideally in a pre-commit hook.
+- **Documentation and resources:** Code quality tools should have robust and helpful documentation, specifically around usage and configuration.
+- **Lean:** Selection of code quality tools should cover all use-cases with little overlap for ease of use and to minimize mental surface area required to understand the system. As in, each tool should have a specific, meaningful purpose.
+- **Enforces security:** Tooling should assist us in identifying security concerns and vulnerabilities.
+- **Well-maintained:** We have a preference towards libraries that are widely adopted and have active maintainers.
+
+## Options Considered
```
Comment (lucasmbrown-usds) [inline documentation/decisions/adr/000N-frontend-tooling.md:22]: do we want to consider something like `immutableJS` to enforce immutability? i know there are lots of pros and cons to that!

**Review by SammySteiner — APPROVED** [review]
This looks great, thanks for putting this together.
I wonder if we can update this in the future as we add more tools.

**Comment by daphnegold** [comment]
> This looks great, thanks for putting this together. I wonder if we can update this in the future as we add more tools.

Definitely, I think that's ideal as we iterate!

File: documentation/decisions/adr/000N-frontend-tooling.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+**[Renovate](https://docs.renovatebot.com/):** Automated dependency updates.
+- Free and open source
+- Can batch, schedule, auto-assign and auto merge updates
+- Multi-language and supports Python as well
+
+*Note: Do we need the Safety package for our Python code as well or is this sufficient?*
+
+### License Checking[^*]
+**[License Checker](https://www.npmjs.com/package/license-checker):** Very aptly named tool to check licenses for dependencies.
```
Comment (daphnegold) [inline documentation/decisions/adr/000N-frontend-tooling.md:62]: Having a bit of trouble finding a license checking tool that will work well with automation. I'll put some more time into it, but let me know if you have any ideas!

File: documentation/decisions/adr/000N-frontend-tooling.md (hunk 1 above)
Comment (daphnegold) [reply documentation/decisions/adr/000N-frontend-tooling.md:22]: Would this be alongside Typescript? Typescript will add immutability functionality that we can use as we desire, and I think that's sufficient for now. I'd need to do a deeper dive into immutable.js to see if it offers something additional. I can add it to the documentation so it is tracked as an option.

**Pull request merged**